src/wrapper/objects/global_ref.rs

```diff
@@ -1,143 +1,96 @@
 use std::convert::From;
-use std::mem;
+use std::sync::Arc;
 
-use JNIEnv;
 use JavaVM;
-use errors::*;
+use JNIEnv;
+use errors::Result;
 use objects::JObject;
-use sys::{
-    self,
-    jobject,
-};
+use sys;
+
 
 /// A global JVM reference. These are "pinned" by the garbage collector and are
 /// guaranteed to not get collected until released. Thus, this is allowed to
-/// outlive the `JNIEnv` that it came from. Still can't cross thread boundaries
-/// since it requires a pointer to the `JNIEnv` to do anything useful with it.
+/// outlive the `JNIEnv` that it came from and can be used in other threads.
+#[derive(Clone)]
 pub struct GlobalRef {
+    inner: Arc<GlobalRefGuard>
+}
+
+
+struct GlobalRefGuard {
     obj: JObject<'static>,
-    env: *mut sys::JNIEnv,
+    vm: JavaVM,
 }
 
+
+unsafe impl Send for GlobalRef {}
+
+
 impl<'a> From<&'a GlobalRef> for JObject<'a> {
     fn from(other: &'a GlobalRef) -> JObject<'a> {
-        other.obj
+        other.as_obj()
     }
 }
 
+
 impl GlobalRef {
-    /// Create a new global reference object. This assumes that
-    /// `CreateGlobalRef` has already been called.
-    pub unsafe fn new(env: *mut sys::JNIEnv, obj: jobject) -> Self {
+    /// Creates a new global reference. This assumes that `NewGlobalRef`
+    /// has already been called.
+    pub unsafe fn new(vm: JavaVM, obj: sys::jobject) -> Self {
```

Comment:
Am I right to assume this method is not supposed to be used by the library clients? If so, is it possible in Rust to make it available only inside the library (= hide from the users)?

Reply:
I agree with this, we can make such methods visible only for this crate itself.